crates/cranelift/src/gc/enabled/drc.rs

```diff
@@ -207,11 +260,244 @@ impl DrcCompiler {
         &self.layouts
     }
 
+    fn alloc_array(
+        &mut self,
+        func_env: &mut FuncEnvironment<'_>,
+        builder: &mut FunctionBuilder<'_>,
+        array_type_index: TypeIndex,
+        init: super::ArrayInit<'_>,
+    ) -> WasmResult<ir::Value> {
+        let interned_type_index = func_env.module.types[array_type_index];
+
+        let len_offset = gc_compiler(func_env)?.layouts().array_length_field_offset();
+        let array_layout = func_env.array_layout(interned_type_index);
+        let base_size = array_layout.base_size;
+        let elem_size = array_layout.elem_size;
+        let align = array_layout.align;
+        let len_to_elems_delta = base_size.checked_sub(len_offset).unwrap();
+
+        // First, compute the array's total size from its base size, element
+        // size, and length.
+        let size = emit_array_size(builder, array_layout, init);
+
+        // Second, now that we have the array object's total size, call the
+        // `gc_alloc_raw` builtin libcall to allocate the array.
+        let array_ref = emit_gc_raw_alloc(
+            func_env,
+            builder,
+            VMGcKind::ArrayRef,
+            interned_type_index,
+            size,
+            align,
+        );
+
+        // Write the array's length into the appropriate slot.
+        let len_addr = func_env.prepare_gc_ref_access(
+            builder,
+            array_ref,
+            Offset::Static(len_offset),
+            BoundsCheck::Object(size),
+        );
+        let len = match init {
+            ArrayInit::Fill { len, .. } => len,
+            ArrayInit::Elems(e) => {
+                let len = u32::try_from(e.len()).unwrap();
+                builder.ins().iconst(ir::types::I32, i64::from(len))
+            }
+        };
+        builder
+            .ins()
+            .store(ir::MemFlags::trusted(), len, len_addr, 0);
+
+        // Compute the address of the first element in the array.
+
+        let len_to_elems_delta = builder
+            .ins()
+            .iconst(ir::types::I64, i64::from(len_to_elems_delta));
+        let mut elem_addr = builder.ins().iadd(len_addr, len_to_elems_delta);
+
+        // Finally, initialize each of the newly-allocated array's elements.
+
+        let array_ty = func_env.types[interned_type_index]
+            .composite_type
+            .unwrap_array();
+        let elem_ty = array_ty.0.element_type;
+
+        let pointer_type = func_env.pointer_type();
+        let elem_size = builder.ins().iconst(pointer_type, i64::from(elem_size));
+
+        match init {
+            ArrayInit::Elems(elems) => {
+                for val in elems {
+                    self.init_field(func_env, builder, elem_addr, &elem_ty, *val)?;
+                    elem_addr = builder.ins().iadd(elem_addr, elem_size);
+                }
+            }
+            ArrayInit::Fill { elem, len: _ } => {
+                // To initialize the array by filling it with `elem`, emit the
```

Comment:
This looks pretty similar to `translate_array_fill` above, would it be possible to refactor the two to share more code?

Reply:
good idea